src/provider/provider_devdax_memory.c

```diff
@@ -0,0 +1,673 @@
+    size_t offset;          // offset of the data
+} devdax_ipc_data_t;
+
+static umf_result_t devdax_get_ipc_handle_size(void *provider, size_t *size) {
+    if (provider == NULL || size == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    *size = sizeof(devdax_ipc_data_t);
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static umf_result_t devdax_get_ipc_handle(void *provider, const void *ptr,
+                                          size_t size, void *providerIpcData) {
+    (void)size; // unused
+
+    if (provider == NULL || ptr == NULL || providerIpcData == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    devdax_memory_provider_t *devdax_provider =
+        (devdax_memory_provider_t *)provider;
+
+    void *value = critnib_get(devdax_provider->fd_offset_map, (uintptr_t)ptr);
+    if (value == NULL) {
+        LOG_ERR("devdax_get_ipc_handle(): getting a value from the IPC cache "
+                "failed (addr=%p)",
+                ptr);
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    devdax_ipc_data_t *devdax_ipc_data = (devdax_ipc_data_t *)providerIpcData;
+    devdax_ipc_data->offset = (size_t)value - 1;
+    strncpy(devdax_ipc_data->dd_path, devdax_provider->path, NAME_MAX - 1);
+    devdax_ipc_data->dd_path[NAME_MAX - 1] = '\0';
+    devdax_ipc_data->dd_size = devdax_provider->size;
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static umf_result_t devdax_put_ipc_handle(void *provider,
+                                          void *providerIpcData) {
+    if (provider == NULL || providerIpcData == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    devdax_memory_provider_t *devdax_provider =
+        (devdax_memory_provider_t *)provider;
+    devdax_ipc_data_t *devdax_ipc_data = (devdax_ipc_data_t *)providerIpcData;
+
+    // verify the path of the /dev/dax
+    if (strncmp(devdax_ipc_data->dd_path, devdax_provider->path, NAME_MAX)) {
+        LOG_ERR("devdax path mismatch (local: %s, ipc: %s)",
+                devdax_provider->path, devdax_ipc_data->dd_path);
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    // verify the size of the /dev/dax
+    if (devdax_ipc_data->dd_size != devdax_provider->size) {
+        LOG_ERR("devdax size mismatch (local: %zu, ipc: %zu)",
+                devdax_provider->size, devdax_ipc_data->dd_size);
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static umf_result_t devdax_open_ipc_handle(void *provider,
+                                           void *providerIpcData, void **ptr) {
+    if (provider == NULL || providerIpcData == NULL || ptr == NULL) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    devdax_memory_provider_t *devdax_provider =
+        (devdax_memory_provider_t *)provider;
+    devdax_ipc_data_t *devdax_ipc_data = (devdax_ipc_data_t *)providerIpcData;
+
+    // verify it is the same devdax
+    // verify the path of the /dev/dax
```

Comment:
are both comment lines here needed? I guess just one should suffice

Reply:
Done